src/abstractdataframe/iteration.jl

```diff
@@ -165,18 +165,12 @@ columns(df::AbstractDataFrame) = eachcol(df, false)
 Base.size(itr::DataFrameColumns) = (size(itr.df, 2),)
 Base.IndexStyle(::Type{<:DataFrameColumns}) = Base.IndexLinear()
 
-@inline function Base.getindex(itr::DataFrameColumns{<:AbstractDataFrame,
-                                                     Pair{Symbol, AbstractVector}},
-                               j::Int)
-    @boundscheck checkbounds(itr, j)
-    _names(itr.df)[j] => itr.df[j]
-end
+Base.@propagate_inbounds Base.getindex(itr::DataFrameColumns{<:AbstractDataFrame,
+                                                             Pair{Symbol, AbstractVector}},
+                                       j::Int) = _names(itr.df)[j] => itr.df[j]
```

Comment:
Maybe keep the `@boundscheck`? It should give a more explicit error than the one coming from `_names(itr.df)[j]`. Then just add `@inbounds` to the next line, and `@inlined` should be enough (instead of `@propagate_inbounds`).

Reply:
OK - I will fix this this way, but let us decide what we do in `similar` and then I will push a fix.